ratelimit_var.go

```diff
@@ -0,0 +1,129 @@
+package goflags
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	stringsutil "github.com/projectdiscovery/utils/strings"
+	timeutil "github.com/projectdiscovery/utils/time"
+)
+
+type RateLimit struct {
+	MaxCount uint
+	Duration time.Duration
+}
+
+type RateLimitMap struct {
+	kv map[string]RateLimit
+}
+
+// Set inserts a value to the map. Format: key=value
+func (rateLimitMap *RateLimitMap) Set(value string) error {
+	if rateLimitMap.kv == nil {
+		rateLimitMap.kv = make(map[string]RateLimit)
+	}
+	var k, v string
+	if idxSep := strings.Index(value, kvSep); idxSep > 0 {
+		k = value[:idxSep]
+		v = value[idxSep+1:]
+	}
+	// note:
+	// - inserting multiple times the same key will override the previous value
+	// - empty string is legitimate value
+
+	if k != "" {
+		rateLimit, err := parseRateLimit(v)
+		if err != nil {
+			return err
+		}
+		rateLimitMap.kv[k] = rateLimit
+	}
+	return nil
+}
+
+// Del removes the specified key
+func (rateLimitMap *RateLimitMap) Del(key string) error {
+	if rateLimitMap.kv == nil {
+		return errors.New("empty runtime map")
+	}
+	delete(rateLimitMap.kv, key)
+	return nil
+}
+
+// IsEmpty specifies if the underlying map is empty
+func (rateLimitMap *RateLimitMap) IsEmpty() bool {
+	return rateLimitMap.kv == nil || len(rateLimitMap.kv) == 0
+}
+
+// AsMap returns the internal map as reference - changes are allowed
+func (rateLimitMap *RateLimitMap) AsMap() map[string]RateLimit {
+	return rateLimitMap.kv
+}
+
+func (rateLimitMap RateLimitMap) String() string {
+	defaultBuilder := &strings.Builder{}
+	defaultBuilder.WriteString("{")
+
+	var items string
+	for k, v := range rateLimitMap.kv {
+		items += fmt.Sprintf("\"%s\"=\"%s\"%s", k, v.Duration.String(), kvSep)
+	}
+	defaultBuilder.WriteString(stringsutil.TrimSuffixAny(items, ",", "="))
+	defaultBuilder.WriteString("}")
+	return defaultBuilder.String()
+}
+
+// RateLimitMapVar adds a ratelimit flag with a longname
+func (flagSet *FlagSet) RateLimitMapVar(field *RateLimitMap, long string, defaultValue []string, usage string) *FlagData {
+	return flagSet.RateLimitMapVarP(field, long, "", defaultValue, usage)
+}
+
+// RateLimitMapVarP adds a ratelimit flag with a short name and long name.
+// It is equivalent to RateLimitMapVar, and also allows specifying ratelimits in days (e.g., "hackertarget=2/d" 2 requests per day, which is equivalent to 24h).
+func (flagSet *FlagSet) RateLimitMapVarP(field *RateLimitMap, long, short string, defaultValue []string, usage string) *FlagData {
+	if field == nil {
+		panic(fmt.Errorf("field cannot be nil for flag -%v", long))
+	}
```

Comment:
I think we should allow csv values in flag value (similar to how we normalize and use in goflags.StringSlice)
currently flag only supports this

```console
./basic -rls "h1=8/2h" -rls "h2=5/10s"
{"h2"="1m50s"="h1"="12h0m0s"}
map[h1:{8 12h0m0s} h2:{5 1m50s}]
```

```console
$ ./basic -rls "h1=8/2h,h2=5/10s"
invalid value "h1=8/2h,h2=5/10s" for flag -rls: parse error: time: unknown unit "h,h" in duration "12h,h2=5"
```